### PR TITLE
chore: test axe examples in ci

### PR DESCRIPTION
**Pull request by scurker**
Get examples in `doc/examples` running on CI:

- [x] ~~Chrome Debugging Protocol~~ _(needs to spin up separate instance of chrome)_
- [x] Jasmine ~~(Existing Issue on CircleCI/Travis [#295](https://github.com/gruntjs/grunt-contrib-jasmine/issues/295))~~ _Swapped to Karma to run Jasmine specs_
- [x] Jest/React
- [x] Mocha
- [x] PhantomJS
- [x] Puppeteer
- [x] QUnit

Added script to run all examples under `doc/examples/*`. Added appveyor config to have examples running on windows. 

## Reviewer checks

**Required fields, to be filled out by PR reviewer(s)**
- [x] Follows the commit message policy, appropriate for next version
- [x] Has documentation updated, a DU ticket, or requires no documentation change
- [x] Includes new tests, or was unnecessary
- [x] Code is reviewed for security by: Stephen

### CODE REVIEW

File: doc/examples/jest_react/package.json [hunk 1]
```diff
@@ -13,7 +13,7 @@
 		"test": "jest"
 	},
 	"devDependencies": {
-		"axe-core": "^3.1.1",
+		"axe-core": "^3.2.2",
```
Comment (stephenmathieson) [inline doc/examples/jest_react/package.json]: Keeping this up-to-date seems tedious. Since we're already refactoring this a bit, would a `*` or `file://` be better?

File: doc/examples/test-examples.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
+const { readdirSync, statSync } = require('fs');
+const { join } = require('path');
+const { spawn } = require('child_process');
+const exampleDirs = readdirSync(__dirname)
+	.map(dir => join(__dirname, dir))
+	.filter(dir => statSync(dir).isDirectory());
+
+async function runner(dir) {
+	let child = spawn('npm install && npm test', {
```
Comment (stephenmathieson) [inline doc/examples/test-examples.js]: IIRC `child_process.spawn` has some gnarly Windows issues. Maybe use [`execa`](https://github.com/sindresorhus/execa) instead? It'd also give us the benefit of removing the below Promise wrapper.

File: doc/examples/test-examples.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+				resolve();
+			} else {
+				reject();
+			}
+		});
+	});
+}
+
+(async () => {
+	await Promise.all(exampleDirs.map(runner))
```
Comment (stephenmathieson) [inline doc/examples/test-examples.js]: The outer closure and this `await` are unnecessary. Just do:

```js
Promise.all(exampleDirs.map(runner)).catch(err => {
  console.error(err)
  process.exit(1)
})
```

File: doc/examples/test-examples.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+		.then(() => {
+			// Return successful exit
+			process.exit();
+		})
+		.catch(err => {
+			console.log(err);
```
Comment (stephenmathieson) [inline doc/examples/test-examples.js]: Errors belong on `stderr`, not `stdout`.

File: doc/examples/test-examples.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
+const { readdirSync, statSync } = require('fs');
+const { join } = require('path');
+const { spawn } = require('child_process');
+const exampleDirs = readdirSync(__dirname)
+	.map(dir => join(__dirname, dir))
+	.filter(dir => statSync(dir).isDirectory());
+
+async function runner(dir) {
```
Comment (stephenmathieson) [inline doc/examples/test-examples.js]: Please remove the `async` keyword here, since `await` is not used in this function.

File: doc/examples/jasmine/karma.conf.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
+// Karma configuration
+module.exports = function(config) {
+	config.set({
+		// base path that will be used to resolve all patterns (eg. files, exclude)
```
Comment (stephenmathieson) [inline doc/examples/jasmine/karma.conf.js]: Can we remove the comments here? IMO they're only helpful for people setting up Karma (you), not for people reading this code (us).

File: appveyor.yml [hunk 7]
```diff
@@ -0,0 +1,20 @@
+version: 1.0.{build}-{branch}
```
Comment (stephenmathieson) [inline appveyor.yml]: Can we make this change separately? Adding AppVeyor should be in its own PR/commit.

File: .circleci/config.yml [hunk 8]
```diff
@@ -92,11 +99,13 @@ workflows:
           requires:
             - dependencies
             - lint
+      - test_examples
```
Comment (stephenmathieson) [inline .circleci/config.yml]: We should wait until the `dependencies`, `lint` and `test` jobs have finished successfully before running the examples.

We want to "fail fast" on CI so we're not blocking other builds unnecessarily.

This should look something like:

```yaml
- test_examples
    requires:
      - test
```

File: doc/examples/jest_react/package.json (hunk 1 above)
Comment (scurker) [reply doc/examples/jest_react/package.json]: It seems since these are primarily intended to be examples to be consumed, neither `*` or `file:...` seem appropriate if we want to promote best practices.

File: doc/examples/test-examples.js (hunk 2 above)
Comment (scurker) [reply doc/examples/test-examples.js]: I made sure there wasn't any windows issues, but it looks like we're using `execa` elsewhere so I can switch to that.

File: doc/examples/test-examples.js (hunk 3 above)
Comment (scurker) [reply doc/examples/test-examples.js]: 👍

File: doc/examples/test-examples.js (hunk 4 above)
Comment (scurker) [reply doc/examples/test-examples.js]: 👍

File: doc/examples/test-examples.js (hunk 5 above)
Comment (scurker) [reply doc/examples/test-examples.js]: 👍

File: .circleci/config.yml (hunk 8 above)
Comment (scurker) [reply .circleci/config.yml]: Good call!

File: appveyor.yml (hunk 7 above)
Comment (scurker) [reply appveyor.yml]: I wanted to ensure that appveyor was running successfully on this pr, but this can be implemented on another pr.

File: doc/examples/jest_react/package.json (hunk 1 above)
Comment (stephenmathieson) [reply doc/examples/jest_react/package.json]: Sounds good.

File: doc/examples/test-examples.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
+const { readdirSync, statSync } = require('fs');
+const { join } = require('path');
+const execa = require('execa');
+const exampleDirs = readdirSync(__dirname)
+	.map(dir => join(__dirname, dir))
+	.filter(dir => statSync(dir).isDirectory());
+
+function runner(dir) {
+	let child = execa('npm install && npm test', {
```
Comment (stephenmathieson) [inline doc/examples/test-examples.js]: `execa` works a bit differently that `child_process.spawn`. This should probably be something like:

```js
function runner(dir) {
  const install = execa('npm', [ 'install' ], { cwd: dir })
  return install.then(()=> execa('npm', ['test'], { cwd: dir })
}
```

File: doc/examples/test-examples.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+		cwd: dir,
+		shell: true,
+		stdio: 'inherit'
+	});
+
+	return new Promise((resolve, reject) => {
```
Comment (stephenmathieson) [inline doc/examples/test-examples.js]: This is no longer necessary since `execa` returns a Promise.

**Comment by stephenmathieson** [comment]
It's probably not a big deal, but it might be nice to cache the dependencies of the examples in CircleCI. This would speed the builds up a bit.

File: doc/examples/test-examples.js (hunk 9 above)
Comment (scurker) [reply doc/examples/test-examples.js]: 👍 That's on me for not looking at the documentation intelligently...

File: doc/examples/test-examples.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+	.map(dir => join(__dirname, dir))
+	.filter(dir => statSync(dir).isDirectory());
+
+async function runner(dir) {
+	const config = { cwd: dir, stdio: 'inherit' };
+
+	let install = await execa.shell('npm install', config);
+	let test = await execa.shell('npm test', config);
+
+	return Promise.all([install, test]);
```
Comment (stephenmathieson) [inline doc/examples/test-examples.js]: Since the `install` and `test` promises have already resolved, we don't need the `Promise.all` here 😄

Comment (scurker) [reply doc/examples/test-examples.js]: Yeah, I realized that after I pushed. Sorry about that!
